src/packages/cli/src/cli.ts

```diff
@@ -38,109 +65,182 @@ const isDocker =
+          err.stack || err
+        ],
+        1
+      );
+    }
+  };
+
+  // See http://stackoverflow.com/questions/10021373/what-is-the-windows-equivalent-of-process-onsigint-in-node-js
+  if (process.platform === "win32") {
+    const rl = (require("readline") as typeof Readline)
+      .createInterface({
+        input: process.stdin,
+        output: process.stdout
+      })
+      .on("SIGINT", () => {
+        // we must "close" the RL interface otherwise the process will think we
+        // are still listening
+        // https://nodejs.org/api/readline.html#readline_event_sigint
+        rl.close();
+        handleSignal("SIGINT");
+      });
   }
-});
 
-let receivedShutdownSignal: boolean = false;
-const handleSignal = async (signal: NodeJS.Signals) => {
-  console.log(`\nReceived shutdown signal: ${signal}`);
-  closeHandler();
-};
-const closeHandler = async () => {
-  try {
-    // graceful shutdown
-    switch (server.status) {
-      case ServerStatus.opening:
-        receivedShutdownSignal = true;
-        console.log("Server is currently starting; waiting…");
-        return;
-      case ServerStatus.open:
-        console.log("Shutting down…");
-        await server.close();
-        console.log("Server has been shut down");
+  process.on("SIGINT", handleSignal);
+  process.on("SIGTERM", handleSignal);
+  process.on("SIGHUP", handleSignal);
+
+  async function startGanache(err: Error) {
+    if (err) {
+      console.error(err);
+      process.exitCode = 1;
+      return;
+    } else if (receivedShutdownSignal) {
+      closeHandler();
+      return;
+    }
+    started = true;
+    switch (flavor) {
+      case FilecoinFlavorName: {
+        await initializeFilecoin(
+          server.provider as FilecoinProvider,
+          cliSettings
+        );
+        break;
+      }
+      case EthereumFlavorName:
+      default: {
+        initializeEthereum(server.provider as EthereumProvider, cliSettings);
         break;
+      }
     }
-    // don't just call `process.exit()` here, as we don't want to hide shutdown
-    // errors behind a forced shutdown. Note: `process.exitCode` doesn't do
-    // anything other than act as a place to anchor this comment :-)
-    process.exitCode = 0;
-  } catch (err: any) {
-    logAndForceExit(
-      [
-        "\nReceived an error while attempting to shut down the server: ",
-        err.stack || err
-      ],
-      1
-    );
-  }
-};
 
-// See http://stackoverflow.com/questions/10021373/what-is-the-windows-equivalent-of-process-onsigint-in-node-js
-if (process.platform === "win32") {
-  const rl = (require("readline") as typeof Readline)
-    .createInterface({
-      input: process.stdin,
-      output: process.stdout
+    // if process.send is defined, this is a child_process (we assume a detached
+    // instance), so we need to notify that we are ready.
+    const isDetachedInstance = process.send !== undefined;
+    if (isDetachedInstance) {
+      notifyDetachedInstanceReady();
```

Comment:
Why are we notifying at this point? The detached instance is _not_ ready at this point, as the server hasn't even been started yet. :thinking:

Reply:
AFAICT the server _is_ started - this is in the callback passed into `server.listen`.

I've refactored this a touch - although I'm not a huge fan of the inline callback, I think this makes the logic flow a little more clear.